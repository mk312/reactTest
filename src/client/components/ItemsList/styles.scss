.result {
  min-height: 35px;
  padding: 14px 75px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #424242;
}

.number {
  display: inline-block;
  font-weight: 800;
}

.sortTitle {
  display: inline-block;
  margin-right: 14px;
}

.list {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
}

.movie {
  flex-basis: 30%;
  flex-grow: 0;
  margin-bottom: 42px;
  font-size: 12px;
  line-height: 21px;
  color: #CBCBCB;
}

.poster {
  display: inline-block;
  overflow: hidden;
  width: 100%;
  margin-bottom: 21px;

  &:hover {
    transition: 0.5s;
    transform: scale(1.05);
  }
}

.title {
  float: left;
  width: 75%;
}

.release {
  float: right;
  padding: 0 7px;
  border: 1px solid #CBCBCB;
  border-radius: 2px;
}

.genre {
  width: 100%;
  overflow: hidden;
}

.error {
  width: 100%;
  padding:75px;
  text-align: center;
  font-size: 21px;
}