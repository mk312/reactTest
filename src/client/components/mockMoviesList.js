const moviesList = {
  data: [
    {
      id: 1,
      title: 'Kill Bill-1',
      tagline: 'Oscar winning',
      vote_average: 5,
      vote_count: 0,
      release_date: '2000',
      poster_path: 'media/51fY8NCyXpL._SX342_.jpg',
      overview: ' American martial arts film written and directed by Quentin Tarantino. '
      + 'It stars Uma Thurman as the Bride, who swears revenge on a team of assassins (Lucy Liu, Michael Madsen, '
      + 'Daryl Hannah, and Vivica A. Fox) and their leader, Bill (David Carradine), after they try to kill her and her '
      + 'unborn child. Her journey takes her to Tokyo, where she battles the yakuza.',
      budget: 0,
      revenue: 0,
      runtime: 154,
      genres: [
        'Drama',
      ],
    },
    {
      id: 2,
      title: 'Kill Bill-2',
      tagline: 'string',
      vote_average: 0,
      vote_count: 0,
      release_date: '2002',
      poster_path: 'media/51fY8NCyXpL._SX342_.jpg',
      overview: 'string',
      budget: 0,
      revenue: 0,
      runtime: 0,
      genres: [
        'Drama',
        'Thriller',
      ],
    },
  ],
};

export default moviesList;
