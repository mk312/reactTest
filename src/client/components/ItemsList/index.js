import styles from './styles.scss';

import React, { PureComponent } from 'react';
import '@babel/polyfill';
import { connect } from 'react-redux';
import { Link } from "react-router-dom";
import { updateSorting, requestMoviesArr, updateMoviesPage } from '../../actions/actions';

import Toggle from '../basic/Toggle/';

class ItemsList extends PureComponent {
    constructor(props) {
        super(props);
        this.handleSortingChange = (event) => {
            this.props.onUpdateSorting(event.target.value);
            this.props.onUpdateMoviesPage(0);
            this.props.onRequestMoviesArr({...this.props.searchParams, offset:0, sortBy: event.target.value});
        };
        this.handleScroll = () =>{
            if(document.body.offsetHeight < window.innerHeight + document.body.scrollTop + 1) {
                let newOffset = this.props.searchParams.offset +1;
                this.props.onUpdateMoviesPage(newOffset);
                this.props.onRequestMoviesArr({...this.props.searchParams, offset: newOffset}, this.props.choosenMovie);
            }
        }
    }

    componentDidMount() {
        this.props.onRequestMoviesArr(this.props.searchParams);
        window.addEventListener('scroll', this.handleScroll);
    }

    render() {
        return (
            <React.Fragment>
                <div className={styles.result}>
                    { this.props.chosenMovie ? (
                            <div className={styles.number}>Films by {this.props.chosenMovie.genres.join(', ')} genre</div>
                        ):
                        (<React.Fragment>
                            <div className={styles.number}>{this.props.moviesList.length || 0} movies found</div>
                            <div className={styles.sort}>
                                <div className={styles.sortTitle}>SORT BY</div>

                                <Toggle handleFilterChange={this.handleSortingChange}
                                        filterValue={this.props.searchParams.sortBy}
                                        values={[{value: 'DATE', id: 'sv1'}, {value: 'RATING', id: 'sv2'}]} />
                            </div>
                        </React.Fragment>)
                    }
                </div>

                <div className={styles.list}>
                    {this.props.moviesList.map((movie) => (
                        <div key={movie.id} className={`js-movie-item ${styles.movie}`}>
                            <Link to={'/movie/'+ movie.id} onClick={() => this.props.handleMovieClick(movie.id)}>
                                <img className={styles.poster} src={movie.poster_path}/>
                            </Link>
                            <div className={styles.title}>{movie.title}</div>
                            <div className={styles.release}>{movie.release_date}</div>
                            <div className={styles.genre}>{movie.genres.join(', ')}</div>
                        </div>
                    ))}
                    { this.props.moviesList.length === 0 ?
                        <div className={styles.error}>No items found</div>
                        : null
                    }
                </div>
            </React.Fragment>
        );
    }
}

export default connect(
    (state) => {
        return {
            moviesList: state.moviesList,
            searchParams: state.searchParams,
            chosenMovie: state.chosenMovie,
        };
    },
    (dispatch) => {
        return {
            onUpdateSorting: (sortBy) => dispatch(updateSorting(sortBy)),
            onUpdateMoviesPage: (offset) => dispatch(updateMoviesPage(offset)),
            onRequestMoviesArr: (searchParams, choosenMovie) => dispatch(requestMoviesArr(searchParams, choosenMovie)),
        }
    },
)(ItemsList);