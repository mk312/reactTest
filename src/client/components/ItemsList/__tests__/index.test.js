import React from 'react';
import {Simulate} from 'react-dom/test-utils';
import {render} from '@testing-library/react'

import ItemsList from '../index';
import { moviesList } from './../../mockMoviesList';

let container = null;

describe('App', () => {
    beforeEach(() => {
        container = render(<ItemsList moviesList = {moviesList}/>);
    });
    it('should toggle between Date/Rating when one of them is clicked', () => {
        const radio = [
            {id:'sv1', value: 'DATE'},
            {id:'sv2', value: 'RATING'},
        ]
        radio.forEach ((item) => {
            const toggleButton = document.getElementById(`${item.id}`);
            Simulate.change(toggleButton, { target: { value: item.value }});

            expect(toggleButton).toHaveProperty("checked");
            // re-check with snapshot
            expect(container).toMatchSnapshot();
        })
    });
});