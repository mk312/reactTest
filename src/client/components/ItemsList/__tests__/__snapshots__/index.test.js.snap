// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should toggle between Date/Rating when one of them is clicked 1`] = `
<div>
  <div
    class="result"
  >
    <div
      class="number"
    >
      2
       movies found
    </div>
    <div
      class="sort"
    >
      <div
        class="sortTitle"
      >
        SORT BY
      </div>
      <input
        checked=""
        class="js-toggle-radio-button"
        id="sv1"
        name="contact"
        type="radio"
        value="DATE"
      />
      <label
        for="sv1"
      >
        DATE
      </label>
      <input
        class="js-toggle-radio-button"
        id="sv2"
        name="contact"
        type="radio"
        value="RATING"
      />
      <label
        for="sv2"
      >
        RATING
      </label>
    </div>
  </div>
  <div
    class="list"
  >
    <div
      class="js-movie-item movie"
    >
      <a
        href="#"
      >
        <img
          class="poster"
          src="media/51fY8NCyXpL._SX342_.jpg"
        />
      </a>
      <div
        class="title"
      >
        Kill Bill-1
      </div>
      <div
        class="release"
      >
        2000
      </div>
      <div
        class="genre"
      >
        Drama
      </div>
    </div>
    <div
      class="js-movie-item movie"
    >
      <a
        href="#"
      >
        <img
          class="poster"
          src="media/51fY8NCyXpL._SX342_.jpg"
        />
      </a>
      <div
        class="title"
      >
        Kill Bill-2
      </div>
      <div
        class="release"
      >
        2002
      </div>
      <div
        class="genre"
      >
        Drama, Thriller
      </div>
    </div>
  </div>
</div>
`;

exports[`App should toggle between Date/Rating when one of them is clicked 2`] = `
<div>
  <div
    class="result"
  >
    <div
      class="number"
    >
      2
       movies found
    </div>
    <div
      class="sort"
    >
      <div
        class="sortTitle"
      >
        SORT BY
      </div>
      <input
        class="js-toggle-radio-button"
        id="sv1"
        name="contact"
        type="radio"
        value="DATE"
      />
      <label
        for="sv1"
      >
        DATE
      </label>
      <input
        checked=""
        class="js-toggle-radio-button"
        id="sv2"
        name="contact"
        type="radio"
        value="RATING"
      />
      <label
        for="sv2"
      >
        RATING
      </label>
    </div>
  </div>
  <div
    class="list"
  >
    <div
      class="js-movie-item movie"
    >
      <a
        href="#"
      >
        <img
          class="poster"
          src="media/51fY8NCyXpL._SX342_.jpg"
        />
      </a>
      <div
        class="title"
      >
        Kill Bill-1
      </div>
      <div
        class="release"
      >
        2000
      </div>
      <div
        class="genre"
      >
        Drama
      </div>
    </div>
    <div
      class="js-movie-item movie"
    >
      <a
        href="#"
      >
        <img
          class="poster"
          src="media/51fY8NCyXpL._SX342_.jpg"
        />
      </a>
      <div
        class="title"
      >
        Kill Bill-2
      </div>
      <div
        class="release"
      >
        2002
      </div>
      <div
        class="genre"
      >
        Drama, Thriller
      </div>
    </div>
  </div>
</div>
`;
